src/openms/source/ANALYSIS/TOPDOWN/DeconvolutedSpectrum.cpp

```diff
@@ -0,0 +1,654 @@
+  {
+    std::vector<PeakGroup>().swap(peakGroups);
+  }
+
+  bool DeconvolutedSpectrum::empty() const
+  {
+    return peakGroups.empty();
+  }
+
+  MSSpectrum DeconvolutedSpectrum::toSpectrum()
+  {
+    auto outSpec = MSSpectrum(*spec);
+    outSpec.clear(false);
+    for (auto &pg : peakGroups)
+    {
+      if (pg.peaks.empty())
+      {
+        continue;
+      }
+      auto p = Peak1D(pg.monoisotopicMass, pg.intensity);
+      outSpec.push_back(p);
+    }
+    if (precursorPeakGroup != nullptr && !spec->getPrecursors().empty())
+    {
+      std::vector<Precursor> precursors;
+      Precursor precursor(spec->getPrecursors()[0]);
+      precursor.setCharge(precursorPeakGroup->maxQScoreCharge);
+      precursor.setMZ(precursorPeakGroup->monoisotopicMass);
+      precursor.setIntensity(precursorPeakGroup->intensity);
+      precursors.push_back(precursor);
+      outSpec.setPrecursors(precursors);
+    }
+
+    return outSpec;
+  }
+
+  void DeconvolutedSpectrum::writeDeconvolutedMasses(std::fstream &fs,
+                                                     FLASHDeconvHelperStructs::Parameter &param)//, fstream &fsm, fstream &fsp)
+  {
+    if (empty())
+    {
+      return;
+    }
+
+    for (auto &pg : peakGroups)
+    {
+      if (pg.peaks.empty())
+      {
+        continue;
+      }
+      double &m = pg.monoisotopicMass;
+      double &am = pg.avgMass;
+      double &intensity = pg.intensity;
+      int minCharge = param.chargeRange + param.minCharge;
+      int maxCharge = -1;
+      for (auto &p : pg.peaks)
+      {
+        minCharge = minCharge < p.charge ? minCharge : p.charge;
+        maxCharge = maxCharge > p.charge ? maxCharge : p.charge;
+      }
+
+      fs << pg.massIndex << "\t" << pg.specIndex << "\t" << param.fileName << "\t" << pg.scanNumber << "\t"
+         << std::to_string(spec->getRT()) << "\t"
+         << peakGroups.size() << "\t"
+         << std::to_string(am) << "\t" << std::to_string(m) << "\t" << intensity << "\t"
+         << minCharge << "\t" << maxCharge << "\t"
+         << pg.peaks.size() << "\t";
+
+      if (param.writeDetail)
+      {
+        fs << std::fixed << std::setprecision(2);
+        for (auto &p : pg.peaks)
+        {
+          fs << p.mz << ";";
+        }
+        fs << "\t";
+
+        fs << std::fixed << std::setprecision(1);
+        for (auto &p : pg.peaks)
+        {
+          fs << p.intensity << ";";
+        }
+        fs << "\t";
+        fs << std::setprecision(-1);
+
+        for (auto &p : pg.peaks)
+        {
+          fs << p.charge << ";";
+        }
+        fs << "\t";
+        for (auto &p : pg.peaks)
+        {
+          fs << p.getUnchargedMass(param.chargeMass) << ";";
+        }
+        fs << "\t";
+        for (auto &p : pg.peaks)
+        {
+          fs << p.isotopeIndex << ";";
+        }
+        fs << "\t";
+
+        for (auto &p : pg.peaks)
+        {
+          auto tm = pg.monoisotopicMass + p.isotopeIndex * Constants::ISOTOPE_MASSDIFF_55K_U;
```

Comment:
comment

Reply:
also line below

Reply:
What's the point of this one? Thanks!

Reply:
I just meant to add a comment on the special constant here

Reply:
Got it!

Reply:
I actually added the comments in Constants.h